data/transactions/logic/assembler.go

```diff
@@ -1485,218 +1483,282 @@ func checkByteConstBlock(cx *evalContext) int {
 	return 1
 }
 
-func disIntcblock(dis *disassembleState, spec *OpSpec) {
-	var intc []uint64
-	intc, dis.nextpc, dis.err = parseIntcblock(dis.program, dis.pc)
-	if dis.err != nil {
-		return
-	}
-	_, dis.err = fmt.Fprintf(dis.out, "intcblock")
-	if dis.err != nil {
-		return
+func disIntcblock(dis *disassembleState, spec *OpSpec) (string, error) {
+	intc, nextpc, err := parseIntcblock(dis.program, dis.pc)
+	if err != nil {
+		return "", err
 	}
+	dis.nextpc = nextpc
+	out := spec.Name
 	for _, iv := range intc {
-		_, dis.err = fmt.Fprintf(dis.out, " %d", iv)
-		if dis.err != nil {
-			return
-		}
+		dis.intc = append(dis.intc, iv)
+		out += fmt.Sprintf(" %d", iv)
 	}
-	_, dis.err = dis.out.Write([]byte("\n"))
+	return out, nil
 }
 
-func disBytecblock(dis *disassembleState, spec *OpSpec) {
-	var bytec [][]byte
-	bytec, dis.nextpc, dis.err = parseBytecBlock(dis.program, dis.pc)
-	if dis.err != nil {
-		return
+func disIntc(dis *disassembleState, spec *OpSpec) (string, error) {
+	lastIdx := dis.pc + spec.Details.Size - 1
+	if len(dis.program) <= lastIdx {
+		missing := lastIdx - len(dis.program) + 1
+		return "", fmt.Errorf("unexpected %s opcode end: missing %d bytes", spec.Name, missing)
 	}
-	_, dis.err = fmt.Fprintf(dis.out, "bytecblock")
-	if dis.err != nil {
-		return
+	dis.nextpc = dis.pc + spec.Details.Size
+	var suffix string
+	var b int
+	switch spec.Opcode {
+	case 0x22:
+		suffix = "_0"
+		b = 0
+	case 0x23:
+		suffix = "_1"
+		b = 1
+	case 0x24:
+		suffix = "_2"
+		b = 2
+	case 0x25:
+		suffix = "_3"
+		b = 3
+	case 0x21:
+		b = int(dis.program[dis.pc+1])
+		suffix = fmt.Sprintf(" %d", b)
+	default:
+		return "", fmt.Errorf("disIntc on %v", spec)
 	}
+	if b < len(dis.intc) {
+		return fmt.Sprintf("intc%s // %d", suffix, dis.intc[b]), nil
+	}
+	return fmt.Sprintf("intc%s", suffix), nil
+}
+
+func disBytecblock(dis *disassembleState, spec *OpSpec) (string, error) {
+	bytec, nextpc, err := parseBytecBlock(dis.program, dis.pc)
+	if err != nil {
+		return "", err
+	}
+	dis.nextpc = nextpc
+	out := spec.Name
 	for _, bv := range bytec {
-		_, dis.err = fmt.Fprintf(dis.out, " 0x%s", hex.EncodeToString(bv))
-		if dis.err != nil {
-			return
+		dis.bytec = append(dis.bytec, bv)
+		out += fmt.Sprintf(" 0x%s", hex.EncodeToString(bv))
+	}
+	return out, nil
+}
+
+func allPrintableASCII(bytes []byte) bool {
+	for _, b := range bytes {
+		if b < 32 || b > 126 {
+			return false
 		}
 	}
-	_, dis.err = dis.out.Write([]byte("\n"))
+	return true
 }
+func guessByteFormat(bytes []byte) string {
+	var short basics.Address
 
-func disPushInt(dis *disassembleState, spec *OpSpec) {
+	if len(bytes) == len(short) {
```

Comment:
this is not 100% accurate, sometimes it can be a random hash that is also 32 bytes long. I had a similar issue in tealdbg with types guessing :(

Reply:
What did you do there?  I figured we want to be as helpful as possible, and there's no strictly correct thing to do, so I would try to present the most likely correct thing.  Nothing goes wrong if it's wrong, but it's nice to get it right.

One thing I realize is that I will certainly reverse the order, so that a 32 byte printable string is printed as a string, if it's entirely printable.  That's vanishingly unlikely for a key.

Reply:
This was for stack/scratch values: I check if printable, and either leave as a string or convert to 0xAABBCC... as assembler/disassembler. I removed address type guessing in a preference of correctness, and wanted to implement tooltip on hover with address string, but switched to something else.

Working solution for untyped addresses would be running the code and assigning `Address` type if a value compared/assigned/retrieved from known addresses (like `txn.Sender` field). But seems like it does not worth efforts :)